### DISCUSSION TITLE
Simple manual for flashing a Faikin

### DISCUSSION THREAD

**Discussion by Y0Sam** [revk/xESP32-Faikout#530] [Show and tell]
I had to flash my devices to upgrade them to the latest firmware as for some reason the OTA update did not work.

Many thanks to revk for all the work he has done to make this possible.

I had to figure out some things that might benefit anyone.

Below a simple manual of what I have done:

Preparation:
To install the ESPtools needed  see the how to: https://docs.espressif.com/projects/esptool/en/latest/esp32/installation.html

Download the correct bin files needed from: 
https://github.com/revk/ESP32-Faikin/tree/main/ESP/release
<your ESP board type>-bootloader.bin e.g Faikin-S3-MINI-N4-R2-bootloader.bin
partition-table.bin
ota_data_initial.bin
<your ESP board type>.bin e.g Faikin-S3-MINI-N4-R2.bin
Place them in a folder/directory of your choice

Connecting the Faikin to your computer:
I used a ESP-WROOM-32 programmer/development board: https://nl.aliexpress.com/item/4000587174844.html?spm=a2g0o.order_list.order_list_main.35.30e979d2DAoLqk&gatewayAdapt=glo2nld
![ESP32-Prog1](https://github.com/user-attachments/assets/fed93312-36ee-4d2b-9aaf-f91fed10fc8c)

For the connection to the Faikin I used a 5 pins  2.54mm pitch single row female PCB pin, as this matched the hole distance on the PCB.header.  For connection between the programmer and the pin header male-female dupont wires are used.
Connect the coloured wires to the pin header and to the pins of the programmer as shown below.
![pinheader](https://github.com/user-attachments/assets/f98cbc9d-fefa-4c3c-85fe-870407264fa1)

![faikin_programmer_wiring](https://github.com/user-attachments/assets/a22c891f-9b1a-4b86-bb78-825a634118b3)

Note that the IO0 pin did not toggle between low and high. 
To be able to program the Faikin the blue wire was directly connected to the GND pin (blue arrows).
Assembled it looked like shown below
![programmer](https://github.com/user-attachments/assets/1de70628-6702-4644-9f10-79a87b2c8665)

The orange wire is GND pin. (Note: in the picture the blue wire is connected to the IO0 pin)
Programming the Faikin:
To program insert the pin header in the holes of the holes of the PCB and hold them in place (as they do not lock in place).
![Faikin](https://github.com/user-attachments/assets/d3d4a237-1f41-4083-90ad-224bb6669d8d)

Open a command window (for windows press windows-key, type CMD and press enter)
Navigate to the folder/directory where you put the downloaded files

Erase Flash:
Copy paste the line below and  in the command window, press enter and wait for the action to finish.
esptool.py --chip esp32s3 erase_flash 

ESPtool.py will automatically try the available com ports until it finds the Faikin and starts erasing the Faikin flash.
The output, when successful will look like shown below:
![erase_output](https://github.com/user-attachments/assets/9a0dca6a-57f7-443f-94a7-419453e6523d)

If encountered an error check if the wire to the IO0 pin of the Faikin is connected to GND.
An other problem to verify is if the programmer is connected and active. This can be checked via the windows device manager. In addition to COM1 there should be another one (of the programmer).
Programming the new firmware:
For this command the COM port to use needs to be know. It will show up in output of the flash erase action (COM7 in the image above). 
Copy paste the applicable line below and  in the command window, enter the correct com port if needed and press enter and wait for the action to finish.
For the S3 chip:
esptool.py -p com7 write_flash 0x0 Faikin-S3-MINI-N4-R2-bootloader.bin 0x8000 partition-table.bin 0xd000 ota_data_initial.bin 0x10000 Faikin-S3-MINI-N4-R2.bin
For the S1 chip:
esptool.py -p /dev/ttyUSB0 write_flash 0x1000 Faikin-S1-PICO-bootloader.bin 0x8000 partition-table.bin 0xd000 ota_data_initial.bin 0x10000 Faikin-S1-PICO.bin

Programming is slow so be patient.
The output should look something like the image below.
![flash_output5](https://github.com/user-attachments/assets/7b97ad10-8613-4b08-ba37-40a481d061d4)

Check that device is flashed correct:
Disconnect the pin header from the Faikin.
Change the IO0 pin on the programmer to a 3v3 or the IO0 pin.
Connect the pin header to the Faikin again.
When the device is flashed successful the led will light up green.
To configure the device see the setup manual: https://github.com/revk/ESP32-Faikin/blob/main/Manuals/Setup.md

**Comment by RaW79** [comment]
Nice but sounds very unnecessarily complicated

Simply use the USB connectivity of the S3 (GND, 5VDC, D+, D-). No special wires, Automatic flashmode etc etc
![grafik](https://github.com/user-attachments/assets/7ed3f15d-a65f-40f0-ad80-54b1ff51e83b)

Flashing is a One-Click-Operation that´s it.
![grafik](https://github.com/user-attachments/assets/414c4ce5-aecf-40cf-8ec2-e32aea372020)

**Reply by revk** [reply]
Agreed, and even the new boards with the TC2030 contact pads are just USB connection (need soldering to pads).

**Reply by Y0Sam** [reply]
I wasn't sure if those were the USB connections, so yes it is more complicated, but it worked ;)
Try his method next time.
Thanx

**Reply by revk** [reply]
Yes, sorry was not clearly USB. Latest boards only have TC2030 USB.

**Reply by RaW79** [reply]
Yes as mentioned soldering might be needed in case of no TC2030 Cable (~40€). Would not say that this is not complicated but you showed the old nice layout for easy&cheap USB connection  :)
![grafik](https://github.com/user-attachments/assets/c1542bc1-51a9-4b23-8927-268afdec197d)

**Reply by dlbogdan** [reply]
quite a pain to flash these without the special cable. still thinking about a best way...

**Reply by revk** [reply]
Sorry, just for production the TC2030 works very well and is a lot easier.

**Reply by dlbogdan** [reply]
I know. I ordered 5 for my own use at jlcpcb (have 3 units in my home) and now I need to figure out how to flash them.

**Reply by RaW79** [reply]
As shown above. [Flux](https://de.aliexpress.com/item/1005006580039229.html?spm=a2g0o.detail.pcDetailBottomMoreOtherSeller.3.71baY7l8Y7l8cI&gps-id=pcDetailBottomMoreOtherSeller&scm=1007.40196.366991.0&scm_id=1007.40196.366991.0&scm-url=1007.40196.366991.0&pvid=58879f88-d316-46e0-b118-ac1ce96880bf&_t=gps-id:pcDetailBottomMoreOtherSeller,scm-url:1007.40196.366991.0,pvid:58879f88-d316-46e0-b118-ac1ce96880bf,tpp_buckets:668%232846%238113%231998&pdp_npi=4%40dis%21EUR%216.56%210.99%21%21%2149.39%217.45%21%4021038da617301240742864503e802d%2112000037725092565%21rec%21DE%21%21ABXZ&utparam-url=scene%3ApcDetailBottomMoreOtherSeller%7Cquery_from%3A) is your friend and be careful not to destroy the contact pads.

**Reply by dlbogdan** [reply]
indeed it is. flashing worked fine after a 5 min soldering extra thin, extra rigid wires. 
![IMG_6605](https://github.com/user-attachments/assets/74397a34-485f-4b54-837a-0060b2e5844e)

looking forward to donate to the devs after I test them in the AC units

**Reply by RaW79** [reply]
Are sure about the connection? Pin1 is unconnected which is VCC. Pin2 is connected which is unused.
Pin 5 and Pin6 seems to be shorted. Only Pin5 used for GND
![grafik](https://github.com/user-attachments/assets/2f0cab86-6dcc-4369-b054-9e475e30c308)

**Reply by revk** [reply]
End pins are pairs

**Reply by aaranstent** [reply]
This might be easier for people who struggle to solder at this size.  https://www.tag-connect.com/product/tc2030-usb-nl

**Reply by dlbogdan** [reply]
Flashed successfully all 5 of the boards, all within the whole of 20 minutes of soldering/unsoldering/flashing. All of that with both my hands broken from a bike crash, my right hand being in cast.

**Reply by aaranstent** [reply]
I must be missing something. When I connect my new 2024-08-16 Faikin board directly to a USB port using the TC2030-USB-NL cable, my PC sees the USB port for 1 second, disconnects, and then reconnects. This repeats over and over. Does this mean I don't have enough power?

**Reply by revk** [reply]
Sounds unlikely - a USB port should have enough power.

**Reply by RaW79** [reply]
@aaranstent had same behavouir at dedicated faikin modules also by using the soldering technic. Others (major part) works fine. Had to conclude that sth is wrong with this modules and could not fix it so far.

**Reply by aaranstent** [reply]
Tried this on a HP PC (rather than usb port on Dell laptop dock) with the same result.  Will start to look at system logs in the morning.  Assume i am not missing any steps?  Just plug it in and it will appear as  USB connected com port ready to flash.  Do I need specific drivers for the device?

**Reply by revk** [reply]
I think we have seen a raspberry pi get upset, but if you catch it with esptool when connected it flashes and is fine. On my Mac it just works. I think we have seen it work better via a power USB hub.

**Reply by RaW79** [reply]
Just had maybe same behavouir today by 3 faikin modules. Using Windows 10 and the device manager is showing a serial device and no device  switching within ~2seconds.  Then stopped wondering because of this and simply connected by the shown above official ESP Flash tool and it´s been connected and successfully flashed. So guess the ESP tool puts the ESP Chip into a dedicated mode (maybe flash mode).

If I am right this is new and different that referenced cases before but worth to give it a try.

**Reply by aaranstent** [reply]
Success.  I was able to force the firmware/flash via the command line.  (Originally I had the N8 before realising I had an N4 chip).

esptool write_flash 0x0 Faikin-S3-MINI-N4-R2-bootloader.bin 0x8000 partition-table.bin 0xd000 ota_data_initial.bin 0x10000 Faikin-S3-MINI-N4-R2.bin

![{F06442A5-73CD-4A8E-B95D-4501B5BCE172}](https://github.com/user-attachments/assets/2a26abab-cc34-4205-897a-913336e19900)

**Reply by dlbogdan** [reply]
all 5 of my blank Faikin pcbs did the same thing with the USB cycling until the actual flashing part took place. It's completely normal.

**Reply by aaranstent** [reply]
It is not clear where the "Boot mode" button is located.  My boards have 2024-08-16 on them.  I cannot see where the GPI00 pin is now located on this updated board.  I wa able to "catch" the other 8 boards I had by running the command above.  At most twice did the trick.

**Reply by revk** [reply]
You do not need boot mode when using USB to flash. Just connect USB and used idf.py to flash.

FYI I have never had problems from a Mac. But as you see just use the device id rather than having it try and find it and maybe a couple of tries.

I also have a flasher module that handles the power on and boot seamlessly. Perhaps I should sell them :-)

**Comment by gentlecat** [comment]
Hey folks. Reusing this discussion to share my experience with flashing Faikin without USB to serial adapter. I got a [broken beta firmware](https://github.com/revk/ESP32-Faikin/issues/588#issuecomment-2563129370) on a couple of my boards and re-flashing seems to be the only way to revive them.

I have a Raspberry Pi around, so I wanted to see if I can use it instead of a USB adapter.

Here are some related pages which I used for reference:

* https://github.com/tasmota/docs-7.1/blob/master/Flash-Sonoff-using-Raspberry-Pi.md
* https://spellfoundry.com/2016/05/29/configuring-gpio-serial-port-raspbian-jessie-including-pi-3-4/
* https://www.raspberrypi.com/documentation/computers/configuration.html#uarts-and-device-tree
* https://raspberrypi.stackexchange.com/a/108781
* https://pinout.xyz

The gist is that some configuration is required to get serial communication working.

One mistake I made initially is not connecting the `GPIO0` pin to anything. That resulted in some failures when using *esptool.py*. Connecting it to ground resolved the problem.

```
~> tio /dev/ttyAMA0
[17:33:47.757] tio v2.5
[17:33:47.757] Press ctrl-t q to quit
[17:33:47.758] Connected
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:2
load:0x3fce3818,len:0x3c8
load:0x403c9700,len:0x4
load:0x403c9704,len:0x9ac
load:0x403cc700,len:0x27b8
entry 0x403c9868
Guru Meditation Error: Core  0 panic'ed (IllegalInstruction). Exception was unhandled.
Memory dump at 0x420074d8: 00007cc7 3040900d a0b20ca0
Core  0 register dump:
PC      : 0x420074dc  PS      : 0x00060d30  A0      : 0x803768d2  A1      : 0x3fceb2a0
A2      : 0x00000000  A3      : 0x00000020  A4      : 0x00000000  A5      : 0xfc000000
A6      : 0x0000003f  A7      : 0x0000000c  A8      : 0x82004b10  A9      : 0x3fceb270
A10     : 0x3fceb2a0  A11     : 0xfffe0000  A12     : 0x00000000  A13     : 0x3fca2cbc
A14     : 0x00000000  A15     : 0x00000005  SAR     : 0x00000006  EXCCAUSE: 0x00000000
EXCVADDR: 0x00000000  LBEG    : 0x40056f5c  LEND    : 0x40056f72  LCOUNT  : 0x00000000


Backtrace: 0x420074d9:0x3fceb2a0 0x403768cf:0x3fceb2d0 0x403cd70e:0x3fceb340 0x403cd95a:0x3fceb380 0x403c98bd:0x3fceb4b0 0x40045c01:0x3fceb570 0x40043ab6:0x3fceb6f0 0x40034c45:0x3fceb710




ELF file SHA256:

Rebooting...
```

To verify that the connection is working without writing anything, I used this command to read the MAC address:

```
~> esptool.py  --chip esp32s3 --port /dev/ttyAMA0 read_mac
esptool.py v4.8.1
Serial port /dev/ttyAMA0
Connecting....
Chip is ESP32-S3 (QFN56) (revision v0.1)
Features: WiFi, BLE, Embedded Flash 4MB (XMC), Embedded PSRAM 2MB (AP_3v3)
Crystal is 40MHz
MAC: ****
Uploading stub...
Running stub...
Stub running...
MAC: f4:12:fa:97:fc:30
Hard resetting via RTS pin...
```

After setting up the connection, it's basically just a matter of using the same process as described in [this guide](https://github.com/revk/ESP32-Faikin/blob/main/Manuals/Setup.md).

```
~> esptool.py --port /dev/ttyAMA0 --chip esp32s3 erase_flash

esptool.py v4.8.1
Serial port /dev/ttyAMA0
Connecting...
Chip is ESP32-S3 (QFN56) (revision v0.1)
Features: WiFi, BLE, Embedded Flash 4MB (XMC), Embedded PSRAM 2MB (AP_3v3)
Crystal is 40MHz
MAC: ****
Stub is already running. No upload is necessary.
Erasing flash (this may take a while)...
Chip erase completed successfully in 16.7s
Hard resetting via RTS pin...
```

```
~/E/E/release (main)> esptool.py --port /dev/ttyAMA0 write_flash 0x0 Faikin-S3-MINI-N4-R2-bootloader.bin 0x8000 partition-table.bin 0xd000 ota_data_initial.bin 0x10000 Faikin-S3-MINI-N4-R2.bin
esptool.py v4.8.1
Serial port /dev/ttyAMA0
Connecting...
Detecting chip type... ESP32-S3
Chip is ESP32-S3 (QFN56) (revision v0.1)
Features: WiFi, BLE, Embedded Flash 4MB (XMC), Embedded PSRAM 2MB (AP_3v3)
Crystal is 40MHz
MAC: ****
Stub is already running. No upload is necessary.
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00003fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000d000 to 0x0000efff...
Flash will be erased from 0x00010000 to 0x00180fff...
Compressed 13888 bytes to 9597...
Wrote 13888 bytes (9597 compressed) at 0x00000000 in 1.0 seconds (effective 106.7 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 194...
Wrote 3072 bytes (194 compressed) at 0x00008000 in 0.1 seconds (effective 393.1 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 31...
Wrote 8192 bytes (31 compressed) at 0x0000d000 in 0.1 seconds (effective 683.9 kbit/s)...
Hash of data verified.
Compressed 1511424 bytes to 931455...
Wrote 1511424 bytes (931455 compressed) at 0x00010000 in 82.6 seconds (effective 146.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

![](https://github.com/user-attachments/assets/861e80a1-0433-49be-9f16-9272929ee3e1)

**Reply by revk** [reply]
For those not wanting to install `esptool`, it can be done with Chrome, see bottom of page https://github.com/revk/ESP32-RevK

**Reply by RaW79** [reply]
Guess there are several of these options available Like: https://github.com/SpacehuhnTech/espwebtool
Which might work but it's really funny according to the topic "**Simple** ...." which might Not fit to every Option but creativity seems to be huge.

**Reply by revk** [reply]
Well I need to investigate the Chrome ESPtool, and see if I can make a page, or a link that is one click somehow.
